Packs/PaloAltoNetworksCortex_v2/Integrations/PaloAltoNetworksCortex_v2/PaloAltoNetworksCortex_v2.py

```diff
@@ -0,0 +1,786 @@
+import base64
+from cryptography.hazmat.primitives.ciphers.aead import AESGCM
+from typing import Dict, Any, List
+import hashlib
+
+# disable insecure warnings
+requests.packages.urllib3.disable_warnings()
+
+''' GLOBAL CONSTS '''
+ACCESS_TOKEN_CONST = 'access_token'
+REFRESH_TOKEN_CONST = 'refresh_token'
+STORED_CONST = 'stored'
+INSTANCE_ID_CONST = 'instance_id'
+API_URL_CONST = 'api_url'
+REGISTRATION_ID_CONST = 'reg_id'
+ENCRYPTION_KEY_CONST = 'auth_key'
+
+''' GLOBAL VARS '''
+REGISTRATION_ID = demisto.params().get(REGISTRATION_ID_CONST)
+# If there's a stored token in integration context, it's newer than current
+REFRESH_TOKEN = demisto.getIntegrationContext().get(REFRESH_TOKEN_CONST) or demisto.params().get(REFRESH_TOKEN_CONST)
+ENC_KEY = demisto.params().get(ENCRYPTION_KEY_CONST)
+USE_SSL = not demisto.params().get('insecure', True)  # TODO set to False
+PROXY = demisto.params().get('proxy', False)
+TOKEN_RETRIEVAL_URL = 'https://oproxy.demisto.ninja'
+FIRST_FETCH_TIMESTAMP = demisto.params().get('first_fetch_timestamp', '24 hours').strip()
+
+
+class Client(BaseClient):
+    """
+    Client will implement the service API, and should not contain any Demisto logic.
+    Should only do requests and return data.
+    """
+
+    def __init__(self):
+        headers = {
+            'Authorization': REGISTRATION_ID,
+            'Accept': 'application/json'
+        }
+        super().__init__(base_url=TOKEN_RETRIEVAL_URL, headers=headers, verify=USE_SSL, proxy=PROXY)
+
+    def get_access_token(self):
+        integration_context = demisto.getIntegrationContext()
+        access_token = integration_context.get(ACCESS_TOKEN_CONST)
+        stored = integration_context.get('stored')
+        if access_token and stored:
+            if int(time.time()) - stored < 60 * 60 - 30:
+                return access_token
+        oproxy_response = self._http_request("POST",
+                                             "/cdl-token",
+                                             json_data={'token': get_encrypted(REFRESH_TOKEN, ENC_KEY)})
+        access_token = oproxy_response.get(ACCESS_TOKEN_CONST)
+        api_url = oproxy_response.get('url')
+        refresh_token = oproxy_response.get(REFRESH_TOKEN_CONST)
+        instance_id = oproxy_response.get(INSTANCE_ID_CONST)
+        updated_integration_context = {
+            ACCESS_TOKEN_CONST: access_token,
+            STORED_CONST: int(time.time()),
+            API_URL_CONST: api_url,
+            INSTANCE_ID_CONST: instance_id
+        }
+        if refresh_token:
+            updated_integration_context.update({REFRESH_TOKEN: refresh_token})
+        demisto.setIntegrationContext(updated_integration_context)
+        return access_token
+
+    def query_loggings(self, query):
+        """
+        This function handles all the querying of Cortex Logging service
+        """
+        query_data = {'query': self.add_instance_id_to_query(query),
+                      'language': 'bigquery'}
```

Comment:
The backend of QueryService is implemented on top of BigQuery but in the future the QS client could support multiple query languages.

Since the query language I'm using now is bigquery thought it's best to keep it stable.